### DISCUSSION TITLE
automerge not working

### DISCUSSION THREAD

**Discussion by ClementParis016** [renovatebot/renovate#12708] [Request Help]
Hi,

On one of our repositories ([Spendesk/terms-and-conditions](https://github.com/Spendesk/terms-and-conditions)), the automerge setting doesn't work and I can't figure out why.

We require at least 1 approval for each PR and all status checks to pass as well as the PR being up-to-date with the base branch. Renovate manages to keep the branches up-to-date but when it runs on the scheduled time it never merges any branches that satisfies those checks.

GitHub settings, status checks and approvals look good:
![Screenshot 2021-11-17 at 11 20 55](https://user-images.githubusercontent.com/5601687/142182922-fdcc2e23-3555-4d05-9088-2a3233a83794.png)

but for the Renovate's job log I'm not sure what's actually significant for debugging:

<details>
<summary>Here is the renovate.json5 content</summary>

```json5
{
  extends: ['config:base'],
  timezone: 'Europe/Paris',
  schedule: ['after 9pm every weekday', 'before 8am every weekday', 'every weekend'],
  labels: ['dependencies'],
  separateMajorMinor: true,
  separateMultipleMajor: true,
  respectLatest: true,
  ignoreUnstable: true,
  rangeStrategy: 'bump',
  enabledManagers: ['npm'],
  ignoreNpmrcFile: true,
  npmrc: '//registry.npmjs.org/:_authToken=${NPM_TOKEN}',
  encrypted: {
    npmToken: '__redacted__',
  },
  commitMessagePrefix: 'chore(env):',
  assigneesFromCodeOwners: true,
  packageRules: [
    {
      matchDepTypes: ['devDependencies'],
      matchPackagePatterns: ['lint', 'prettier'],
      automerge: true,
    },
    {
      matchUpdateTypes: ['minor', 'patch', 'pin', 'digest'],
      automerge: true,
    },
  ],
  rebaseWhen: 'never',
}
```

</details>

<details>
<summary>And here an extract of the job log 509058284 for one of the dependencies that is never merged by Renovate</summary>

```txt
DEBUG: Setting current branch to master(branch="renovate/marked-2.x")
DEBUG: latest commit(branch="renovate/marked-2.x")
{
  "branchName": "master",
  "latestCommitDate": "2021-11-16T10:22:03+01:00"
}
DEBUG: branchExists=true(branch="renovate/marked-2.x")
DEBUG: dependencyDashboardCheck=undefined(branch="renovate/marked-2.x")
DEBUG: PR rebase requested=false(branch="renovate/marked-2.x")
DEBUG: Checking if PR has been edited(branch="renovate/marked-2.x")
DEBUG: Found existing branch PR(branch="renovate/marked-2.x")
DEBUG: Checking schedule(after 9pm every weekday,before 8am every weekday,every weekend, Europe/Paris)(branch="renovate/marked-2.x")
DEBUG: Found timezone(branch="renovate/marked-2.x")
{
  "timezone": "Europe/Paris"
}
DEBUG: Adjusting now for timezone(branch="renovate/marked-2.x")
DEBUG: Checking 3 schedule(s)(branch="renovate/marked-2.x")
DEBUG: Checking schedule "after 9pm every weekday"(branch="renovate/marked-2.x")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_a": [
          75600
        ],
        "d": [
          2,
          3,
          4,
          5,
          6
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}
DEBUG: Checking schedule "before 8am every weekday"(branch="renovate/marked-2.x")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_b": [
          28800
        ],
        "d": [
          2,
          3,
          4,
          5,
          6
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}
DEBUG: Matches schedule before 8am every weekday(branch="renovate/marked-2.x")
DEBUG: Branch already exists(branch="renovate/marked-2.x")
DEBUG: Skipping stale branch check due to rebaseWhen=never(branch="renovate/marked-2.x")
DEBUG: Branch does not need rebasing(branch="renovate/marked-2.x")
DEBUG: Using reuseExistingBranch: true(branch="renovate/marked-2.x")
DEBUG: manager.getUpdatedPackageFiles() reuseExistinbranch=true(branch="renovate/marked-2.x")
DEBUG: npm.updateDependency(): devDependencies.@types/marked = ^2.0.5(branch="renovate/marked-2.x")
DEBUG: No package files need updating(branch="renovate/marked-2.x")
DEBUG: Getting updated lock files(branch="renovate/marked-2.x")
DEBUG: Writing package.json files(branch="renovate/marked-2.x")
{
  "packageFiles": [
    "package.json"
  ]
}
DEBUG: Writing package-lock.json(branch="renovate/marked-2.x")
DEBUG: Massaging npm lock file before writing to disk(branch="renovate/marked-2.x")
DEBUG: Writing any updated package files(branch="renovate/marked-2.x")
DEBUG: No updated lock files in branch(branch="renovate/marked-2.x")
DEBUG: Skipping commit (rebaseWhen=never)(branch="renovate/marked-2.x")
```

</details>

In the job log there are a few things that may give clues, for ex. "Skipping commit" (?) but I can't tell what is relevant or not in this.

Could you point me in the right direction? I've been tweaking settings for the past few days without success :/

**Comment by viceice** [comment]
- Renovate can't skip the approval.
- Renovate uses PR's for automerge by default, so PR without review is never ready to merge for renovate.

**Reply by rarkins** [reply]
Does it need Renovate approve bot?

**Reply by ClementParis016** [reply]
@viceice yes I'm aware of that, I don't expect it to merge without prior approval, but the problem is even with approval it still doesn't merge anything.

For the PR I'm talking about in the job log, it's been approved & green since weeks:
![Screenshot 2021-11-17 at 11 20 55](https://user-images.githubusercontent.com/5601687/142182792-b699afc3-409c-4d9d-b57d-6ca052a44511.png)

**Reply by rarkins** [reply]
Can you take a screenshot of your branch protection rules and paste it here?

**Reply by ClementParis016** [reply]
@rarkins  sure, here you go:
![github com_Spendesk_terms-and-conditions_settings_branch_protection_rules_21609882 (1)](https://user-images.githubusercontent.com/5601687/142220552-aac1bed2-8510-4d77-b481-046fc86a06ec.png)

btw, we have an other repository with the exact same settings (it seems, unless I missed something), except for "Require review from Code Owners", but where Renovate does its job normally and automerges correctly

**Reply by rarkins** [reply]
IIRC if there is a codeowner for the file(s) being updated then GitHub won't allow the merge from a non-admin until the codeowner has approved. It's a somewhat undocumented feature/limitation

**Reply by ClementParis016** [reply]
In this case the codeowner is a GitHub team and only people from that team have access to the repository (so only them can approve) so it seems that wouldn't be the issue? I'll try to toggle it off before the next scheduled run to see if it changes anything

**Reply by rarkins** [reply]
The issue here is that GitHub won't allow Renovate to merge, and that seems to be because a codeowner hasn't approved. It's not because Renovate is choosing not to merge.

**Reply by ClementParis016** [reply]
That still doesn't work even without the requirement on code owners approval 🤔 

<details>
<summary>Extract of job log 509527215 for branch="renovate/marked-2.x"</summary>

```txt
DEBUG: Setting current branch to master(branch="renovate/marked-2.x")
DEBUG: latest commit(branch="renovate/marked-2.x")
{
  "branchName": "master",
  "latestCommitDate": "2021-11-16T10:22:03+01:00"
}
DEBUG: branchExists=true(branch="renovate/marked-2.x")
DEBUG: dependencyDashboardCheck=undefined(branch="renovate/marked-2.x")
DEBUG: PR rebase requested=false(branch="renovate/marked-2.x")
DEBUG: Checking if PR has been edited(branch="renovate/marked-2.x")
DEBUG: Found existing branch PR(branch="renovate/marked-2.x")
DEBUG: Checking schedule(after 9pm every weekday,before 8am every weekday,every weekend, Europe/Paris)(branch="renovate/marked-2.x")
DEBUG: Found timezone(branch="renovate/marked-2.x")
{
  "timezone": "Europe/Paris"
}
DEBUG: Adjusting now for timezone(branch="renovate/marked-2.x")
DEBUG: Checking 3 schedule(s)(branch="renovate/marked-2.x")
DEBUG: Checking schedule "after 9pm every weekday"(branch="renovate/marked-2.x")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_a": [
          75600
        ],
        "d": [
          2,
          3,
          4,
          5,
          6
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}
DEBUG: Matches schedule after 9pm every weekday(branch="renovate/marked-2.x")
DEBUG: Branch already exists(branch="renovate/marked-2.x")
DEBUG: Skipping stale branch check due to rebaseWhen=never(branch="renovate/marked-2.x")
DEBUG: Branch does not need rebasing(branch="renovate/marked-2.x")
DEBUG: Using reuseExistingBranch: true(branch="renovate/marked-2.x")
DEBUG: manager.getUpdatedPackageFiles() reuseExistinbranch=true(branch="renovate/marked-2.x")
DEBUG: npm.updateDependency(): devDependencies.@types/marked = ^2.0.5(branch="renovate/marked-2.x")
DEBUG: No package files need updating(branch="renovate/marked-2.x")
DEBUG: Getting updated lock files(branch="renovate/marked-2.x")
DEBUG: Writing package.json files(branch="renovate/marked-2.x")
{
  "packageFiles": [
    "package.json"
  ]
}
DEBUG: Writing package-lock.json(branch="renovate/marked-2.x")
DEBUG: Massaging npm lock file before writing to disk(branch="renovate/marked-2.x")
DEBUG: Writing any updated package files(branch="renovate/marked-2.x")
DEBUG: No updated lock files in branch(branch="renovate/marked-2.x")
DEBUG: Skipping commit (rebaseWhen=never)(branch="renovate/marked-2.x")
```

</details>

**Reply by rarkins** [reply]
You need a new branch to be created to test it, and also that log seems to end too soon. Finally, why are you using rebaseWhen=never with automerge anyway?

**Reply by ClementParis016** [reply]
Ok I removed the `rebaseWhen: 'never'` and it finally worked!

We were using it because it rebases too aggressively, as soon as a PR is outdated from the base branch it rebases directly without following the schedule, which is very annoying when you have lots of Renovate PRs because it rebases each opened PR each time it merges one.

Is there any reason why it respects the schedule only to create & merge PRs but not to update them?

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#updatenotscheduled

**Reply by ClementParis016** [reply]
Thanks for your help, I now have a satisfying setup.

I think the problem was coming from a mix of `rebaseWhen: 'never'`, codeowners, and GitHub settings regarding keeping branches up-to-date and required codeowners approval.

<details>
<summary>View final renovate.json5</summary>

```json5
{
  extends: ['config:base'],
  timezone: 'Europe/Paris',
  schedule: ['after 9pm every weekday', 'before 8am every weekday', 'every weekend'],
  updateNotScheduled: 'false',
  labels: ['dependencies'],
  separateMajorMinor: true,
  separateMultipleMajor: true,
  respectLatest: true,
  ignoreUnstable: true,
  rangeStrategy: 'bump',
  enabledManagers: ['npm'],
  ignoreNpmrcFile: true,
  npmrc: '//registry.npmjs.org/:_authToken=${NPM_TOKEN}',
  encrypted: {
    npmToken: '__redacted__',
  },
  commitMessagePrefix: 'chore(env):',
  assigneesFromCodeOwners: true,
  automerge: true,
  packageRules: [
    {
      matchDepTypes: ['devDependencies'],
      matchPackagePatterns: ['lint', 'prettier'],
      automergeType: 'branch',
    },
    {
      matchUpdateTypes: ['minor', 'patch', 'pin', 'digest'],
      automergeType: 'branch',
    },
  ],
}

```

</details>

<details>
<summary>View final GitHub branch protection rules</summary>

![github com_Spendesk_terms-and-conditions_settings_branch_protection_rules_21609882](https://user-images.githubusercontent.com/5601687/142832463-269a6a63-5abe-48c8-b41c-583101669b9a.png)

</details>
